cmd/cluster/core/destroy.go

```diff
@@ -159,23 +157,118 @@ func DestroyCluster(ctx context.Context, hostedCluster *hyperv1.HostedCluster, o
+		if err := c.Get(clusterDeleteCtx, types.NamespacedName{Namespace: o.Namespace, Name: o.Name}, hostedCluster); err != nil {
+			if apierrors.IsNotFound(err) {
+				return true, nil
+			}
+			o.Log.Error(err, "Failed to get hosted cluster", "namespace", o.Namespace, "name", o.Name)
+			return false, nil
+		}
+		done := len(hostedCluster.Finalizers) == 1 && hostedCluster.Finalizers[0] == destroyFinalizer
+		return done, nil
+	}, clusterDeleteCtx.Done())
+
+	if err != nil {
+		return fmt.Errorf("hostedcluster wasn't finalized, aborting delete: %w", err)
+	}
+	return nil
+}
+
+func setFinalizer(ctx context.Context, hostedCluster *hyperv1.HostedCluster, o *DestroyOptions, c client.Client) error {
+	if sets.NewString(hostedCluster.Finalizers...).Has(destroyFinalizer) {
+		return nil
+	}
+
+	original := hostedCluster.DeepCopy()
+	if hostedCluster.DeletionTimestamp == nil {
+		controllerutil.AddFinalizer(hostedCluster, destroyFinalizer)
+	}
+	if o.DestroyCloudResources {
+		if hostedCluster.Annotations == nil {
+			hostedCluster.Annotations = map[string]string{}
+		}
+		hostedCluster.Annotations[hyperv1.CleanupCloudResourcesAnnotation] = "true"
+	}
+	if err := c.Patch(ctx, hostedCluster, client.MergeFrom(original)); err != nil {
+		if apierrors.IsNotFound(err) {
+			o.Log.Info("Hosted cluster not found, skipping finalizer update", "namespace", o.Namespace, "name", o.Name)
+		} else if !strings.Contains(err.Error(), "no new finalizers can be added if the object is being deleted") {
+			return fmt.Errorf("failed to add finalizer to hosted cluster: %w", err)
+		}
+	} else {
+		o.Log.Info("Updated finalizer for hosted cluster", "namespace", o.Namespace, "name", o.Name)
 	}
 
-	o.Log.Info("Successfully destroyed cluster and infrastructure", "namespace", o.Namespace, "name", o.Name, "infraID", o.InfraID)
 	return nil
 }
 
-func DestroyPlatformSpecificsNoop(_ context.Context, _ *DestroyOptions) error {
+func waitForClusterDeletion(ctx context.Context, hostedCluster *hyperv1.HostedCluster, o *DestroyOptions, c client.Client) error {
+	clusterDeleteCtx, clusterDeleteCtxCancel := context.WithTimeout(ctx, o.ClusterGracePeriod)
+	defer clusterDeleteCtxCancel()
+
+	err := wait.PollImmediateUntil(1*time.Second, func() (bool, error) {
+		if err := c.Get(clusterDeleteCtx, types.NamespacedName{Namespace: o.Namespace, Name: o.Name}, hostedCluster); err != nil {
+			if apierrors.IsNotFound(err) {
+				return true, nil
+			}
+			o.Log.Error(err, "Failed to get hosted cluster", "namespace", o.Namespace, "name", o.Name)
+			return false, nil
+		}
+
+		// don't wait for grace period. Nothing happens after grace period in the controller, but it's only
+		// for debug. So it's safe to continue in case of grace period.
+		if _, ok := hostedCluster.Annotations[hyperv1.CleanupCloudResourcesAnnotation]; ok {
+			if meta.FindStatusCondition(hostedCluster.Status.Conditions, string(hyperv1.HostedClusterDestroyed)) != nil {
+				return true, nil
+			}
```

Comment:
does this need the same fix that you provided in https://github.com/openshift/hypershift/pull/3289?